_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+**entire** filepath of the file, and those can get quite lengthy. In my local
+build, `__FILE__` is 77 bytes!
+
+```
+(gdb) p __FILE__
+$1 = "/Users/tyler/dev/memfault/interrupt/example/debugging-asserts/impls/assert3.c"
+```
+
+Thankfully, the `__FILE__` string table entry is re-used between asserts in the
+same file, so the code size bloat primarily depends on the number of files with
+asserts within them. As a rough estimate, if we have asserts in 50 different
+files, this results in **~50 bytes \* 50 files => 2500 bytes**, which is an
+enormous amount of bloat just for assert strings.
+
+For this reason, many RTOS implementations that include an `ASSERT` macro can
+optionally leave these strings out of the build, such as Zephyr[^8] and
+MyNewt[^9].
+
+> NOTE: One simple code size (and sanity) optimization is to normalize these
+> `__FILE__` paths to the root of the project directory. This will result in a
+> build path that removes the `/Users/tyler/dev/memfault/...` portion. An added
+> benefit of this is that all all machines should produce builds with the same
+> filepaths, resulting in more similarly sized builds.
+
+#### Extracted Information
+
+We know the file and line number of the assert hit.
+
+### File**name** and Line Number Assert (#4)
+
+Since the filepaths can be lengthy and waste code space, one solution is to use
+only the filename itself, e.g. `assert3.c`, instead of the filepath, e.g.
+`/Users/tyler/.../assert3.c`.
+
+This can be accomplished by adding a define called `__FILENAME__` and using this
+in the assert implementation instead. In our Makefile or similar build system,
+this is defined along with the other `CFLAGS` as follows.
+
+```
+CFLAGS += -D__FILENAME__=\"$(notdir $<)\"
+```
+
+In the assert header, we now use `__FILENAME__` instead of `__FILE__`.
+
+```c
+#define MY_ASSERT(expr)                   \
+  do {                                    \
+    if (!(expr)) {                        \
+      my_assert(__FILENAME__, __LINE__);  \
+    }                                     \
+  } while (0)
+```
+
+With this change, we reduce our file string from
+`/Users/tyler/dev/memfault/interrupt/example/debugging-asserts/impls/assert3.c`
+to `assert3.c`, resulting in 68 bytes of code space savings per file which
+contains an assert.
+
+```
+$ arm-none-eabi-objdump -s -j '.text' build/nrf52.elf
+ ...
+26b0 61737365 7274342e 6300               assert4.c
+```
+
+#### Extracted Information
+
+We know the file and line number of the assert hit.
+
+### Register Values Only Assert (#5)
+
+The problem with all of the above assert implementations is that our code size
+is bloated due to storing string filenames, and also that our code size can
+change depending on how we name the files in our project. This encourages poor
+naming practices, as well as produces unexpected build sizes.
+
+My favorite way to implement asserts to record both the Program Counter and the
```

Comment:
Consistency: program counter and link register weren't previously capitalized

I think a word is also missing: `asserts is to`